### ISSUE TITLE
emsg in file and mpd

### ISSUE THREAD

**Issue by RufaelDev** [Dash-Industry-Forum/DASH-IF-IOP#233]
Dear All, the emsg box defined in dash has a message payload unsigned int(8)   message_data[]; when carried in ISOBMFF 23009-1. When carried in mpd for some schemes like scte xml, xml+bin, and bin (see also https://github.com/Dash-Industry-Forum/DASH-IF-IOP/issues/199), some processing is needed with part of the event message data being base 64 encoded and part of it being xml  while being stored as a string in the mpd, this is quite confusing when converting from the ISOBMFF to mpd and back, or even when generating an mpd as the way the binary message is presented depends on the scheme. What would make things easier was if the message_data was coded independently using base64, which is a straightforward way of putting the binary content in the mpd without breaking the xml scheme or comprising data integrity, hence we propose supporting a base64 type field for the messageData attribute in the event type as a way to reduce this ambiguity. In addition more explanation of the current sting based approach would be helpful.

**Comment by RufaelDev** [comment]
One solution would be the handline in azure proposed by microsoft:

https://docs.microsoft.com/en-us/azure/media-services/previous/media-services-specifications-live-timed-metadata#22-fragmented-mp4-ingest-smooth-streaming

!-- Example Section in MPD -->

<EventStream schemeIdUri=”urn:example:signaling:1.0” timescale=”1000” value=”player-statistics”>
  <Event presentationTime=”0” duration=”10000” id=”0”> PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0idXRmLTgiPz48QWNxdWlyZWRTaWduYWwgeG1sbnM9InVybjpjYWJsZWxhYnM6bWQ6eHNkOnNpZ25hbGluZzozLjAiIGFjcXVpc2l0aW9uUG9pbnRJZGVudGl0eT0iRVNQTl9FYXN0X0FjcXVpc2l0aW9uX1BvaW50XzEiIGFjcXVpc2l0aW9uU2lnbmFsSUQ9IjRBNkE5NEVFLTYyRkExMUUxQjFDQTg4MkY0ODI0MDE5QiIgYWNxdWlzaXRpb25UaW1lPSIyMDEyLTA5LTE4VDEwOjE0OjI2WiI+PFVUQ1BvaW50IHV0Y1BvaW50PSIyMDEyLTA5LTE4VDEwOjE0OjM0WiIvPjxTQ1RFMzVQb2ludERlc2NyaXB0b3Igc3BsaWNlQ29tbWFuZFR5cGU9IjUiPjxTcGxpY2VJbnNlcnQgc3BsaWNlRXZlbnRJRD0iMzQ0NTY4NjkxIiBvdXRPZk5ldHdvcmtJbmRpY2F0b3I9InRydWUiIHVuaXF1ZVByb2dyYW1JRD0iNTUzNTUiIGR1cmF0aW9uPSJQVDFNMFMiIGF2YWlsTnVtPSIxIiBhdmFpbHNFeHBlY3RlZD0iMTAiLz48L1NDVEUzNVBvaW50RGVzY3JpcHRvcj48U3RyZWFtVGltZXM+PFN0cmVhbVRpbWUgdGltZVR5cGU9IkhTUyIgdGltZVZhbHVlPSI1MTUwMDAwMDAwMDAiLz48L1N0cmVhbVRpbWVzPjwvQWNxdWlyZWRTaWduYWw+</Event>
  <Event presentationTime=”20000” duration=”10000” id=”1”> PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0idXRmLTgiPz48QWNxdWlyZWRTaWduYWwgeG1sbnM9InVybjpjYWJsZWxhYnM6bWQ6eHNkOnNpZ25hbGluZzozLjAiIGFjcXVpc2l0aW9uUG9pbnRJZGVudGl0eT0iRVNQTl9FYXN0X0FjcXVpc2l0aW9uX1BvaW50XzEiIGFjcXVpc2l0aW9uU2lnbmFsSUQ9IjRBNkE5NEVFLTYyRkExMUUxQjFDQTg4MkY0ODI0MDE5QiIgYWNxdWlzaXRpb25UaW1lPSIyMDEyLTA5LTE4VDEwOjE0OjI2WiI+PFVUQ1BvaW50IHV0Y1BvaW50PSIyMDEyLTA5LTE4VDEwOjE0OjM0WiIvPjxTQ1RFMzVQb2ludERlc2NyaXB0b3Igc3BsaWNlQ29tbWFuZFR5cGU9IjUiPjxTcGxpY2VJbnNlcnQgc3BsaWNlRXZlbnRJRD0iMzQ0NTY4NjkxIiBvdXRPZk5ldHdvcmtJbmRpY2F0b3I9InRydWUiIHVuaXF1ZVByb2dyYW1JRD0iNTUzNTUiIGR1cmF0aW9uPSJQVDFNMFMiIGF2YWlsTnVtPSIxIiBhdmFpbHNFeHBlY3RlZD0iMTAiLz48L1NDVEUzNVBvaW50RGVzY3JpcHRvcj48U3RyZWFtVGltZXM+PFN0cmVhbVRpbWUgdGltZVR5cGU9IkhTUyIgdGltZVZhbHVlPSI1MTYyMDAwMDAwMDAiLz48L1N0cmVhbVRpbWVzPjwvQWNxdWlyZWRTaWduYWw+</Event>
</EventStream>

**Comment by ojw28** [comment]
I agree it would be better if the message data were coded using base64 (or similar). From a player perspective, I'm not really sure if we're doing it right in ExoPlayer, but we currently have some pretty awkward code that tries to read the XML corresponding to the event from the XML parser we're using to read the manifest, and serialize the event back into binary form:

https://github.com/google/ExoPlayer/blob/r2.9.2/library/dash/src/main/java/com/google/android/exoplayer2/source/dash/manifest/DashManifestParser.java#L911

It's probably wrong. At best it's very error prone. If the message data were coded using base64 then I think it would be trivial.

**Comment by RufaelDev** [comment]
Thanks for the pointer that code looks complicated for what should be trivial, we try to avoid embedding xml in xml from a different namespace when we can, for us it was also hard to convert from file to mpd correctly. I will discuss it more in the event TF 7am pacific 14 dec, and develop the solution a bit more, eventually this needs to be reported back to MPEG
